### PR TITLE
don't emit a source map without -g4. refactor debug info logic

### PR DESCRIPTION
**Pull request by kripken**
* Refactor the debug info logic to get rid of some really weird conditions.
 * Add a test for only emitting a source map when requested.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -2404,7 +2396,12 @@ def check_bad_eq(arg):
       requested_level = newargs[i][2:] or '3'
       options.debug_level = validate_arg_level(requested_level, 4, 'Invalid debug level: ' + newargs[i])
       options.requested_debug = newargs[i]
-      newargs[i] = ''
+      if options.debug_level >= 3:
+        # make sure debug info is preserved (note: clang won't accept -g4)
+        newargs[i] = '-g'
+      else:
+        # don't pass -g1 or -g2 to clang
+        newargs[i] = ''
```
Comment (sbc100) [inline emcc.py]: Why not?

File: tests/test_other.py [hunk 2]
```diff
@@ -2216,7 +2216,12 @@ def test_debuginfo(self):
           (['-O2', '-g'], True),
         ]:
         print(args, expect_debug)
+        self.clear()
         err = run_process([PYTHON, EMCC, path_from_root('tests', 'hello_world.cpp')] + args, stdout=PIPE, stderr=PIPE).stderr
+        if '-g4' in args:
+          assert os.path.exists('a.out.wasm.map'), 'a source map should be emitted'
```
Comment (sbc100) [inline tests/test_other.py]: I think we have self.assertExists() and self.assertNotExists() these days don't we?

File: tests/test_other.py [hunk 3]
```diff
@@ -5140,8 +5145,8 @@ def test(opts):
         run_process([PYTHON, EMCC, path_from_root('tests', 'hello_world.c'), '-g' + str(i), '-o', 'a' + str(i) + '.bc'] + opts)
         sizes[i] = os.path.getsize('a' + str(i) + '.bc')
       print('  ', sizes)
-      assert sizes['_'] == sizes[0] == sizes[1] == sizes[2] == sizes[3], 'no debug or <4 debug, means no llvm debug info ' + str(sizes)
-      assert sizes['g'] == sizes[4], '-g or -g4 means llvm debug info ' + str(sizes)
+      assert sizes['_'] == sizes[0] == sizes[1] == sizes[2], 'no debug or <3 debug, means no llvm debug info ' + str(sizes)
```
Comment (sbc100) [inline tests/test_other.py]: This seems like oldd behaviour but I guess its the existing behaviour..

File: emcc.py (hunk 1 above)
Comment (kripken) [reply emcc.py]: I'm not sure what it does in clang, but we've never passed it, so I'm not sure it's worth the risk of a change here.

**Comment by kripken** [comment]
Tests found that this PR was wrong before. In fastcomp, we use LLVM debug info to issue various warnings, like nan canonicalization etc. For those to work, we must keep debug info there, even if there isn't a source map later which makes it seem like it isn't needed.

Added a commit to fix that, hopefully things should be ok now.

**Pull request closed without merging**